modules/core/04-channel/types/msgs.go

```diff
@@ -517,18 +517,97 @@ var _ sdk.Msg = &MsgChannelUpgradeTry{}
 
 // NewMsgChannelUpgradeTry constructs a new MsgChannelUpgradeTry
 // nolint:interfacer
-func NewMsgChannelUpgradeTry() *MsgChannelUpgradeTry {
-	return &MsgChannelUpgradeTry{}
+func NewMsgChannelUpgradeTry(
+	portID, channelID string,
+	counterpartyChannel Channel,
+	counterpartySequence uint64,
+	proposedUpgradeChannel Channel,
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	proofChannel []byte,
+	proofUpgradeTimeout []byte,
+	proofUpgradeSequence []byte,
+	proofHeight clienttypes.Height,
+	signer string,
+) *MsgChannelUpgradeTry {
+	return &MsgChannelUpgradeTry{
+		PortId:                 portID,
+		ChannelId:              channelID,
+		CounterpartyChannel:    counterpartyChannel,
+		CounterpartySequence:   counterpartySequence,
+		ProposedUpgradeChannel: proposedUpgradeChannel,
+		TimeoutHeight:          timeoutHeight,
+		TimeoutTimestamp:       timeoutTimestamp,
+		ProofChannel:           proofChannel,
+		ProofUpgradeTimeout:    proofUpgradeTimeout,
+		ProofUpgradeSequence:   proofUpgradeSequence,
+		ProofHeight:            proofHeight,
+		Signer:                 signer,
+	}
 }
 
 // ValidateBasic implements sdk.Msg
 func (msg MsgChannelUpgradeTry) ValidateBasic() error {
-	return errors.New("error method not implemented")
+	if err := host.PortIdentifierValidator(msg.PortId); err != nil {
+		return sdkerrors.Wrap(err, "invalid port ID")
+	}
+
+	if !IsValidChannelID(msg.ChannelId) {
+		return ErrInvalidChannelIdentifier
+	}
+
+	if msg.CounterpartyChannel.State != INITUPGRADE {
+		return sdkerrors.Wrapf(ErrInvalidChannelState, "expected: %s, got: %s", INITUPGRADE, msg.CounterpartyChannel.State)
+	}
+
+	if msg.CounterpartySequence == 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidSequence, "counterparty sequence cannot be 0")
+	}
+
+	if msg.ProposedUpgradeChannel.State != TRYUPGRADE {
+		return sdkerrors.Wrapf(ErrInvalidChannelState, "expected: %s, got: %s", TRYUPGRADE, msg.CounterpartyChannel.State)
+	}
+
+	if msg.TimeoutHeight.IsZero() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidHeight, "timeout height must be non-zero")
+	}
+
+	if msg.TimeoutTimestamp == 0 {
```

Comment:
the spec has both timeout timestamp and timeout height, but do both of these need to have a value or just one?

Reply:
I see this part of pseudocode in the spec in the `chanUpgradeTry` function:

```golang
// either timeout height or timestamp must be non-zero
// if the upgrade feature is implemented on the TRY chain, then a relayer may submit a TRY transaction after the timeout.
// this will restore the channel on the executing chain and allow counterparty to use the ChannelUpgradeCancelMsg to restore their channel.
if timeoutHeight == 0 && timeoutTimestamp == 0 {
    restoreChannel()
    return
}
```

which would indicate that both can be zero. 

Would the relayer submit both zero timeouts in a try message after a timeout was processed on the side that initialized the upgrade to restore the original channel on the counterparty? @AdityaSripal

Reply:
Although on `chanOpenTry` there's a verification that the init side has stored the upgrade timeout:

```golang
abortTransactionUnless(verifyChannelUpgradeTimeout(connection, proofHeight, proofUpgradeTimeout, currentChannel.counterpartyPortIdentifier, currentChannel.counterpartyChannelIdentifier, upgradeTimeout)) 
```

And on `chanOpenInit` there's a condition to abort unless one of the timeouts is zero:

```
// either timeout height or timestamp must be non-zero
abortTransactionUnless(counterpartyTimeoutHeight != 0 || counterpartyTimeoutTimestamp != 0)
```

So how would it be possible to reach this check on `chanUpgradeTry` then?

```golang
// either timeout height or timestamp must be non-zero
// if the upgrade feature is implemented on the TRY chain, then a relayer may submit a TRY transaction after the timeout.
// this will restore the channel on the executing chain and allow counterparty to use the ChannelUpgradeCancelMsg to restore their channel.
if timeoutHeight == 0 && timeoutTimestamp == 0 {
    restoreChannel()
    return
}
```

Reply:
```
So how would it be possible to reach this check on chanUpgradeTry then?
```

@crodriguezvega I guess technically the source chain could have a different/forked implementation for the `UPGRADE_INIT` step.